File: P01/06.hs
recr :: (a -> [a] -> b -> b) -> b -> [a] -> b
recr _ z [] = z
recr f z (x : xs) = f x xs (recr f z xs)

-- a.
sacarUna :: Eq a => a -> [a] -> [a]
sacarUna x = recr f []
    where
        f y ys r
            | x == y    = ys
            | otherwise = y : r
 
-- b.
{-
No es adecuado ya que no podemos 'frenar' prematuramente la evaluación.
foldr, foldl aplican f a todos los elementos. Se podrían usar, igual, si
se mantiene un 'contexto' conjunto al acumulado.
-}

-- c.
insertarOrdenado :: Ord a => a -> [a] -> [a]
insertarOrdenado x = recr f [x]
    where
        f y ys r
            | x < y     = x : y : ys
            | otherwise = y : r

File: P08/06.pl
% aplanar(+Xs, -Ys), Ys contiene los elementos de todos los niveles de Xs.
aplanar([], []).
aplanar([[] | Xs], Ys) :- aplanar(Xs, Ys).
aplanar([[Z | Zs] | Xs], Ys) :- aplanar([Z | Zs], Fs), aplanar(Xs, Rs), append(Fs, Rs, Ys).
aplanar([X | Xs], [X | Ys]) :- X \= [], X \= [_|_], aplanar(Xs, Ys).
